src/analysis/processing/qgsalgorithmkmeansclustering.cpp

```diff
@@ -156,21 +174,18 @@ QVariantMap QgsKMeansClusteringAlgorithm::processAlgorithm( const QVariantMap &p
 
     feedback->setProgress( 50 + i * step );
     QgsAttributes attr = feat.attributes();
-    if ( !feat.hasGeometry() )
+    auto obj = idToObj.find( feat.id() );
+    if ( !feat.hasGeometry() || obj == idToObj.end() )
     {
-      attr << QVariant();
+      attr << QVariant() << QVariant();
     }
     else if ( k <= 1 )
     {
-      attr << 0;
-    }
-    else if ( !idToObj.contains( feat.id() ) )
-    {
-      attr << QVariant();
+      attr << 0 << featureWithGeometryCount;
     }
     else
     {
-      attr << clusterFeatures[ idToObj.value( feat.id() ) ].cluster;
+      attr << clusterFeatures[ idToObj[ feat.id() ] ].cluster << clusterSize[ clusterFeatures[ idToObj[ feat.id() ] ].cluster ];
```

Comment:
avoids the multiple lookups, and simplifies the code:

```suggestion
     int cluster = clusterFeatures[ *obj ].cluster;
      attr << cluster << clusterSize[ cluster ];
```